marconi-chain-index/test-compare-cardano-db-sync/Utxo.hs

```diff
@@ -0,0 +1,429 @@
+          )
+        . C.deserialiseFromRawBytes (C.AsHash C.AsScriptData)
+
+instance SQLite.FromRow DbSyncUtxo where
+  fromRow =
+    DbSyncUtxo
+      <$> SQLite.field
+      <*> SQLite.field
+      <*> SQLite.field
+      <*> SQLite.field
+      <*> SQLite.field
+      <*> SQLite.field
+      <*> SQLite.field
+
+instance PG.FromField Word64 where
+  fromField f meta = fromIntegral @Integer <$> PG.fromField f meta
+
+instance PG.FromField C.SlotNo where
+  fromField f meta = fromIntegral @Integer <$> PG.fromField f meta
+
+instance PG.ToField C.SlotNo where
+  toField = PG.toField . coerce @C.SlotNo @Word64
+
+instance SQLite.ToRow DbSyncUtxo where
+  toRow u =
+    [ SQLite.toField (dbTxId u)
+    , SQLite.toField (dbTxIndex u)
+    , SQLite.toField (dbAddress u)
+    , SQLite.toField (dbSlotNo u)
+    , SQLite.toField (dbBlockHash u)
+    , SQLite.toField (dbBlockNo u)
+    , SQLite.toField (dbDatumHash u)
+    ]
+
+instance PG.FromRow DbSyncUtxo where
+  fromRow =
+    DbSyncUtxo
+      <$> PG.field
+      <*> PG.field
+      <*> PG.field
+      <*> PG.field
+      <*> PG.field
+      <*> PG.field
+      <*> PG.field
+
+deriving newtype instance PG.ToRow C.SlotNo
+
+-- * create a streaming source/destination to stream a supperset of the utxo_view from cardano-db-sync to marconi-sqlite
+
+-- | create a streaming source
+mkStreamingSource
+  :: (MonadIO m)
+  => ((r -> IO ()) -> IO ())
+  -> Stream (Of r) m ()
+mkStreamingSource action =
+  let chanSource
+        :: MonadIO m
+        => TMChan r
+        -> Stream (Of r) m ()
+      chanSource chan = loop
+        where
+          loop = do
+            p <- liftIO $ atomically $ readTMChan chan
+            case p of
+              Just x -> S.yield x >> loop
+              Nothing -> pure ()
+   in do
+        chan <- liftIO newTMChanIO
+        _ <- liftIO $
+          forkIO $ do
+            action $ atomically . writeTMChan chan
+            void (liftIO $ atomically $ closeTMChan chan)
+        chanSource chan
+
+-- | Stream rows from cardano-db-sync, no param substition
+sourceQuery_
+  :: (MonadIO m, PG.FromRow r)
+  => PG.Connection
+  -> PG.Query
+  -> Stream (Of r) m ()
+sourceQuery_ conn q = mkStreamingSource $ PG.forEach_ conn q
+
+-- | Stream rows from cardano-db-sync, with param substition
+sourceQuery
+  :: (PG.ToRow params, PG.FromRow r, MonadIO m)
+  => PG.Connection
+  -> PG.Query
+  -> params
+  -> Stream (Of r) m ()
+sourceQuery conn q params = mkStreamingSource $ PG.forEach conn q params
+
+data DbSyncUtxo = DbSyncUtxo
+  { dbTxId :: C.TxId
+  , dbTxIndex :: Word64
+  , dbAddress :: C.AddressAny
+  , dbSlotNo :: C.SlotNo
+  , dbBlockHash :: C.Hash C.BlockHeader
+  , dbBlockNo :: Word64
+  , dbDatumHash :: !(Maybe (C.Hash C.ScriptData))
+  }
+  deriving (Eq, Show)
+
+-- | print the list of offending addresses. These are addresses in db-sync and not in marconi that have utxos
```

Comment:
@berewt the discrepancies are stored in the delta table in sqlite.
If there are any discrepancies, they are also printed